// apollo-server/dist/integrations/hapiApollo.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const graphql = require('graphql');
const GraphiQL = require('../modules/renderGraphiQL');
const runQuery_1 = require('../core/runQuery');
class ApolloHAPI {
    constructor() {
        this.register = (server, options, next) => {
            server.route({
                method: 'POST',
                path: '/',
                handler: (request, reply) => __awaiter(this, void 0, void 0, function* () {
                    let optionsObject;
                    if (isOptionsFunction(options)) {
                        try {
                            optionsObject = yield options(request);
                        }
                        catch (e) {
                            reply(`Invalid options provided to ApolloServer: ${e.message}`).code(500);
                        }
                    }
                    else {
                        optionsObject = options;
                    }
                    if (!request.payload) {
                        reply('POST body missing.').code(500);
                        return;
                    }
                    const responses = yield processQuery(request.payload, optionsObject);
                    if (responses.length > 1) {
                        reply(responses);
                    }
                    else {
                        const gqlResponse = responses[0];
                        if (gqlResponse.errors && typeof gqlResponse.data === 'undefined') {
                            reply(gqlResponse).code(400);
                        }
                        else {
                            reply(gqlResponse);
                        }
                    }
                }),
            });
            next();
        };
        this.register.attributes = {
            name: 'graphql',
            version: '0.0.1',
        };
    }
}
exports.ApolloHAPI = ApolloHAPI;
class GraphiQLHAPI {
    constructor() {
        this.register = (server, options, next) => {
            server.route({
                method: 'GET',
                path: '/',
                handler: (request, reply) => {
                    const q = request.query || {};
                    const query = q.query || '';
                    const variables = q.variables || '{}';
                    const operationName = q.operationName || '';
                    const graphiQLString = GraphiQL.renderGraphiQL({
                        endpointURL: options.endpointURL,
                        query: query || options.query,
                        variables: JSON.parse(variables) || options.variables,
                        operationName: operationName || options.operationName,
                    });
                    reply(graphiQLString).header('Content-Type', 'text/html');
                },
            });
            next();
        };
        this.register.attributes = {
            name: 'graphiql',
            version: '0.0.1',
        };
    }
}
exports.GraphiQLHAPI = GraphiQLHAPI;
function processQuery(body, optionsObject) {
    return __awaiter(this, void 0, void 0, function* () {
        const formatErrorFn = optionsObject.formatError || graphql.formatError;
        let isBatch = true;
        if (!Array.isArray(body)) {
            isBatch = false;
            body = [body];
        }
        let responses = [];
        for (let payload of body) {
            try {
                const operationName = payload.operationName;
                let variables = payload.variables;
                if (typeof variables === 'string') {
                    variables = JSON.parse(variables);
                }
                let params = {
                    schema: optionsObject.schema,
                    query: payload.query,
                    variables: variables,
                    rootValue: optionsObject.rootValue,
                    context: optionsObject.context,
                    operationName: operationName,
                    logFunction: optionsObject.logFunction,
                    validationRules: optionsObject.validationRules,
                    formatError: formatErrorFn,
                    formatResponse: optionsObject.formatResponse,
                };
                if (optionsObject.formatParams) {
                    params = optionsObject.formatParams(params);
                }
                responses.push(yield runQuery_1.runQuery(params));
            }
            catch (e) {
                responses.push({ errors: [formatErrorFn(e)] });
            }
        }
        return responses;
    });
}
function isOptionsFunction(arg) {
    return typeof arg === 'function';
}
